src/actions/app.cr

```diff
@@ -49,86 +50,139 @@ module DppmRestApi::Actions::App
+    app.service.{{action.id}}
   end
+
+  {% end %}
   # get the status of the service associated with the given application
   relative_get "/:app_name/service/status" do |context|
+    app_name = context.params.url["app_name"]
     raise Unauthorized.new context unless Actions.has_access? context, Access::Read
     # TODO: get the status of the service
   end
-  # stop the service associated with the given application
-  relative_put "/:app_name/service/stop" do |context|
-    raise Unauthorized.new context unless Actions.has_access? context, Access::Update
-    # TODO: stop the service
-  end
+
   # lists dependent library packages
   relative_get "/:app_name/libs" do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Read
     # TODO: list dependencies
   end
-  # return the base application package
-  relative_get "/:app_name/app" do |context|
-    raise Unauthorized.new context unless Actions.has_access? context, Access::Read
-    # TODO: return the base application package
+
+  def valid_streams_for(app : DPPM::Prefix::App)
+    Array(String).new.tap do |streams|
+      app.each_log_stream { |stream| streams << stream }
+    end
   end
-  # returns information present in pkg.con as JSON
-  relative_get "/:app_name/pkg" do |context|
+
+  relative_get "/:app_name/valid-log-streams" do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Read
-    # TODO: return package data
+    build_json context.response do |builder|
+      builder.array do
+        Actions
+          .prefix
+          .new_app(context.params.url["app_name"])
+          .each_log_stream { |stream| builder.string stream }
+      end
+    end
   end
+
   # if the `"stream"` query parameter is set, attempt to upgrade to a websocket
   # and stream the results. Otherwise return a JSON-formatted output of the
   # current log data.
   relative_get "/:app_name/logs" do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Read
-    # TODO: upgrade to websocket or output logs to date
+    app_name = context.params.url["app_name"]
+    app = Actions.prefix.new_app app_name
+    valid_streams = valid_streams_for app
+    stream_names = context.params.query.fetch_all("stream_type") || valid_streams
+    # filter irrellevant values
+    stream_names &= valid_streams
+    if context.params.query["stream"]?
+      raise BadRequest.new context, "\
+        You must use separate requests to stream more than one log stream. A \
+        stream can be selected using the query parameter stream_type." if stream_type.size != 1
+      context.response.upgrade do |socket|
+        app.get_logs(
+          stream_type.first,
+          follow: true,
+          lines: context.params.query["lines"]?.try &.to_i
+        ) { |line| socket.puts line }
+      end
+    else
+      JSON.build context.response do |builder|
+        builder.object do
+          builder.field "logs" do
```

Comment:
Adding a field `logs` isn't needed, we already know we will get logs because of the endpoint.